### PR TITLE
perf: eliminate double string concat in remote-task-runner shutdown logging

### PR DESCRIPTION
**Pull request by jasonk000**
### Description

Improve the performance of `TaskQueue::manage` on large installations by removing unnecessary String concat / construction.

Once the changes in #12096 are applied, the performance of the `TaskQueue` loop becomes much tighter, and is dominated by the logging calls.

Screenshot of profiler showing `TaskQueue-Mana...` thread:
image

Screenshot of profile flamegraph for this thread, highlighting the `format` calls in the stack:
image

During the task clean up loop, the `shutdown()` call is issued multiple times to the `RemoteTaskRunner` - note especially it uses the three-argument invocation of `shutdown()` method:

https://github.com/apache/druid/blob/476d0bf4be4199e97695bd568d165cda98523d37/indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java#L328-L335

This hits the default implementation, which constructs a "reason" argument and passes it on to the two-argument `shutdown()`:

https://github.com/apache/druid/blob/476d0bf4be4199e97695bd568d165cda98523d37/indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunner.java#L91-L94

In particular on large clusters with thousands of tasks, this `default void shutdown` call performs a `StringUtils.format` on a task set with thousands of task IDs, and it serialises all of them to a String.

Even in the event this log line is turned off, the log is still constructed, and then discarded.

### Key changed/added classes in this PR

This introduces an `@Override public void shutdown()` with arguments, and has it perform the log construction and issue only if the info level is enabled. This results in significantly lower CPU consumption in this loop.

This follows up the changes in #12096, and also follows the mailing list discussion here:
https://lists.apache.org/thread/9jgdwrodwsfcg98so6kzfhdmn95gzyrj

This PR has:
- [x] been self-reviewed.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] been tested in a test Druid cluster (as part of a larger block of changes).

### CODE REVIEW

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/RemoteTaskRunner.java [hunk 1]
```diff
@@ -553,11 +553,33 @@ public boolean isWorkerRunningTask(ZkWorker worker, String taskId)
    * Finds the worker running the task and forwards the shutdown signal to the worker.
    *
    * @param taskId - task id to shutdown
+   * @param reasonFormat - a StringUtils reason string for shutdown
+   * @param reasonArgs - args will be used during constructio of reason log line
+   */
+  @Override
+  public void shutdown(final String taskId, String reasonFormat, Object... reasonArgs)
+  {
+    if (log.isInfoEnabled()) {
```
Comment (FrankChen021) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/RemoteTaskRunner.java:562]: no need to check if isInfoEnabled, because the `log.info` internally checks it.

**Comment by FrankChen021** [comment]
Since the `reason` is passed to `shutdown` only for logging, why do we do what you have changed in the interface so that all subclasses gain this improvement?

Something ike this:

```
Logger getLogger();

default void shutdown(String taskid, String reasonFormat, Object... args) 
 { 
   getLogger().info("Shutdown [%s] because: [" + reasonFormat + "]" + reasonFormat, taskid, args);
   shutdown(taskid); 
 } 
```


Another thing that confuses me is that the flamechart shows that there's some bottlenecks at the logging, not the `StringUtils.format`, so to gain the performance improvement, the right thing here to do is to change the logging level to WARN or ERROR right? What you solved here is just to eliminate some string construction under WARN or ERROR logging level, right? 

In another word, at INFO logging level, the gain seems not significant.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/RemoteTaskRunner.java (hunk 1 above)
Comment (jasonk000) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/RemoteTaskRunner.java:562]: Checking explicitly avoids the extra `+` string concatenation in the case that info() is not enabled. This is minor, but while we are here...

**Comment by jasonk000** [comment]
I like your suggestion to move it up to the interface level.

WRT warn / error / etc. You're right; there's two perspectives on this:

1. Yes, the biggest benefit is if logging is set to WARN/ERROR. In my stack, with thousands of tasks running, I specifically set warn level for this class in log4j xml as it is noisy, expensive and provides little benefit for regular operations, but this benefit is mitigated with the current codebase.

2. Even in the case of info logging being enabled, the triple string concat is unnecessary.

**Comment by jasonk000** [comment]
Some extra data, ..

With reference to the TaskQueue-Manager thread, a capture with info enabled:
image

With respect to this thread only, the `TaskRunner::shutdown` call is 82% of CPU in the `manage()` loop, and, the useful part of the loop, `findWorkerRunningTask` consumes ~18%, the rest is logging related. By (1) turning off INFO logging on this path, and (2) avoiding the `StringUtils.format()` call, we get approx 4x more capacity through the `TaskQueue-Manager` thread.
image

In other words, if you take the image just above ^, and eliminate the `Logger::info=` and `StringUtils::format` paths, it opens a lot of capacity through this loop.

This becomes more obvious when coupled to #12096 and the yet-to-be-raised PR to reduce contention on the TaskQueue loop.

**Comment by jasonk000** [comment]
Capturing some test case results from #12099 we have more concrete example:

Alternating between these two lines:
```
  default void shutdown(String taskid, String reasonFormat, Object... reasonArgs)
  {
    // String reason = (getLogger().isInfoEnabled()) ? StringUtils.format(reasonFormat, reasonArgs) : null;
    String reason = StringUtils.format(reasonFormat, reasonArgs);
    shutdown(taskid, reason);
  }
```
The existing approach to always format the message completes the TaskQueueScaleTest in 15.x seconds:
```
[INFO] Running org.apache.druid.indexing.overlord.TaskQueueScaleTest
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 15.553 s - in org.apache.druid.indexing.overlord.TaskQueueScaleTest
```
Using the new way to only format if necessary completes the test in 6.x seconds:
```
[INFO] Running org.apache.druid.indexing.overlord.TaskQueueScaleTest
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 6.529 s - in org.apache.druid.indexing.overlord.TaskQueueScaleTest
```

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunner.java [hunk 2]
```diff
@@ -88,11 +89,17 @@
    */
   void shutdown(String taskid, String reason);
 
-  default void shutdown(String taskid, String reasonFormat, Object... args)
+  default void shutdown(String taskid, String reasonFormat, Object... reasonArgs)
   {
-    shutdown(taskid, StringUtils.format(reasonFormat, args));
+    String reason = (getLogger().isInfoEnabled()) ? StringUtils.format(reasonFormat, reasonArgs) : null;
```
Comment (FrankChen021) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunner.java:94]: better to add some comments here to elaborate why we do this here

Comment (jasonk000) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunner.java:96]: fixed in 7c563eac03 and fb3cf82664

**Review by FrankChen021 — APPROVED** [review]

**Comment by jasonk000** [comment]
Hi @FrankChen021 I moved the detailed log output to debug instead of info; I think this is a better balance. Let me know your thoughts.

In 5685eafeb4.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunner.java [hunk 3]
```diff
@@ -88,11 +89,19 @@
...
+    // only calculate the 'reason' string for debug level logging
+    // in large clusters the 'reasonArgs' may be very large / expensive if it includes a list of tasks
+    String reason = (getLogger().isDebugEnabled()) ? StringUtils.format(reasonFormat, reasonArgs) : "debug log disabled";
+    shutdown(taskid, reason);
   }
 
+  /**
+   * Get the logger. Not expected to be called by consumers.
+   */
+  Logger getLogger();
```
Comment (samarthjain) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunner.java:103]: I would probably change the default implementation to throw an `UnsupportedOperationException`. This will also have the benefit of not having to change classes implementing `TaskRunner` interface where they don't have a logger available (like test classes).

File: indexing-service/src/test/java/org/apache/druid/indexing/overlord/TaskRunnerTest.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+  /**
+   * EasyMock does not support mocking of toString, so this provides a custom
+   * object implementation to track whether toString was called.
+   */
+  public static class ToStringMock
+  {
+    AtomicBoolean wasCalled = new AtomicBoolean(false);
+
+    public boolean getWasCalled()
```
Comment (samarthjain) [inline indexing-service/src/test/java/org/apache/druid/indexing/overlord/TaskRunnerTest.java:79]: nit: maybe a better method name would be `wasCalled()` ?

File: indexing-service/src/test/java/org/apache/druid/indexing/overlord/TaskRunnerTest.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    taskRunner.shutdown("taskid", "reason %s", new Object[]{expensiveObject});
+
+    Assert.assertFalse(expensiveObject.getWasCalled());
+  }
+
+  /**
+   * EasyMock does not support mocking of toString, so this provides a custom
+   * object implementation to track whether toString was called.
+   */
+  public static class ToStringMock
```
Comment (samarthjain) [inline indexing-service/src/test/java/org/apache/druid/indexing/overlord/TaskRunnerTest.java:75]: nit: We should probably have these inner classes `ToStringMock` and `MockTaskRunner` as private static.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunner.java [hunk 6]
```diff
@@ -88,11 +89,19 @@
    */
   void shutdown(String taskid, String reason);
 
-  default void shutdown(String taskid, String reasonFormat, Object... args)
+  default void shutdown(String taskid, String reasonFormat, Object... reasonArgs)
   {
-    shutdown(taskid, StringUtils.format(reasonFormat, args));
+    // only calculate the 'reason' string for debug level logging
+    // in large clusters the 'reasonArgs' may be very large / expensive if it includes a list of tasks
+    String reason = (getLogger().isDebugEnabled()) ? StringUtils.format(reasonFormat, reasonArgs) : "debug log disabled";
```
Comment (samarthjain) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunner.java:96]: If debug logging is not enabled, then the following log line won't be reported. 
https://github.com/apache/druid/blob/master/indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java#L582

Comment (samarthjain) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunner.java:96]: Similarly, this - 
https://github.com/apache/druid/blob/master/indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java#L299

Comment (samarthjain) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunner.java:96]: Looking at the callers who pass in taskIds, I see only TaskQueue#shutdown() that is passing in the task ids. So, I propose we modify code in TaskQueue#shutdown something like this:
```
// Kill tasks that shouldn't be running
    final Set<String> knownTaskIds = tasks
        .stream()
        .map(Task::getId)
        .collect(Collectors.toSet());
    final Set<String> tasksToKill = Sets.difference(runnerTaskFutures.keySet(), knownTaskIds);
    if (!tasksToKill.isEmpty()) {
      log.info("Asking taskRunner to clean up %,d tasks.", tasksToKill.size());
      // On large installations running several thousands of tasks,
      // concatenating the list of known task ids can be compupationally expensive.
      boolean logKnownTaskIds = log.isDebugEnabled();
      String reason = logKnownTaskIds
                      ? String.format("Task is not in knownTaskIds[%s]", knownTaskIds)
                      : "Task is not in knownTaskIds";
      for (final String taskId : tasksToKill) {
        try {
          taskRunner.shutdown(
              taskId, reason
          );
        }
        catch (Exception e) {
          log.warn(e, "TaskRunner failed to clean up task: %s", taskId);
        }
      }
    }
```

If we do this, then you won't have to add a new `getLogger()` method in `TaskRunner` interface.

Comment (jasonk000) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunner.java:96]: I like this solution, I'll swap over to this implementation.

Comment (jasonk000) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskRunner.java]: thanks @samarthjain , this solution is much simpler - fixes applied

let me know if all good and i can squash it

**Comment by samarthjain** [comment]
+1, looks good. I will merge after CI completes.

**Comment by samarthjain** [comment]
Looking at the CI failures, it looks like the build timed out because `HashPartitionMultiPhaseParallelIndexingTest` never completed. To be double sure, I ran the test locally and it completed fine. I am going to go ahead and merge this PR since it looks like tests have been flaky in general off late and I don't think logging changes introduced by this PR can contribute to the failures.

**Comment by clintropolis** [comment]
this PR has broken the ci, forbidden api check is failing, `String.format` shouldn't be called directly, use `StringUtils.format` instead

**Comment by samarthjain** [comment]
Looks like the last rebase you pushed reverted your String.format changes, @jasonk000 . I will submit a PR to fix CI, @clintropolis

**Comment by samarthjain** [comment]
Submitted https://github.com/apache/druid/pull/12304 to fix CI.

**Comment by jasonk000** [comment]
Oops my bad, i thought we had that fixed last night. Sorry!
